airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/state_converters/datetime_stream_state_converter.py

```diff
@@ -113,23 +137,36 @@ def convert_to_sequential_state(self, cursor_field: CursorField, stream_state: M
         """
         if self.is_state_message_compatible(stream_state):
             legacy_state = stream_state.get("legacy", {})
-            if slices := stream_state.pop("slices", None):
-                latest_complete_time = self._get_latest_complete_time(slices)
-                if latest_complete_time:
-                    legacy_state.update({cursor_field.cursor_field_key: self.output_format(latest_complete_time)})
+            latest_complete_time = self._get_latest_complete_time(stream_state["low_water_mark"], stream_state.get("slices", []))
+            if latest_complete_time is not None:
+                legacy_state.update({cursor_field.cursor_field_key: self.output_format(latest_complete_time)})
             return legacy_state or {}
         else:
             return stream_state
 
-    def _get_latest_complete_time(self, slices: List[MutableMapping[str, Any]]) -> Optional[datetime]:
+    def _get_latest_complete_time(self, previous_sync_end: datetime, slices: List[MutableMapping[str, Any]]) -> Optional[datetime]:
         """
         Get the latest time before which all records have been processed.
         """
         if slices:
-            first_interval = self.merge_intervals(slices)[0][self.END_KEY]
-            return first_interval
+            merged_intervals = self.merge_intervals(slices)
+            first_interval = merged_intervals[0]
+            if previous_sync_end < first_interval[self.START_KEY]:
+                # There is a region between `previous_sync_end` and the first interval that hasn't been synced yet, so
+                # we don't advance the state message timestamp
+                return previous_sync_end
+
+            if first_interval[self.START_KEY] <= previous_sync_end <= first_interval[self.END_KEY]:
+                # `previous_sync_end` is between the beginning and end of the first interval, so we know we've synced
+                # up to `self.END_KEY`
+                return first_interval[self.END_KEY]
+
+            # `previous_sync_end` falls outside of the first interval; this is unexpected because we shouldn't have tried
+            # to sync anything before `previous_sync_end`, but we can handle it anyway.
+            return self._get_latest_complete_time(previous_sync_end, merged_intervals[1:])
```

Comment:
This would mean that the first interval is completely before the first interval. In which case can this occur?

Reply:
Like the comment mentions, we don't really expect this to happen. Since it's unexpected I can modify this to raise an exception instead.